libraries/tock-register-interface/src/macros.rs

```diff
@@ -49,14 +57,82 @@ macro_rules! register_bitmasks {
                               $values); )*
     };
     {
-        $valtype:ty, $reg_desc:ident, $(#[$outer:meta])* $field:ident,
+        $valtype:ident, $reg_desc:ident, $(#[$outer:meta])* $field:ident,
+                    $offset:expr, $numbits:expr,
+                    [$( $(#[$inner:meta])* $valname:ident = $value:expr ),+]
+    } => { // this match arm is duplicated below with an allowance for 0 elements in the valname -> value array,
+        // to seperately support the case of zero-variant enums not supporting non-default
+        // representations.
+        #[allow(non_upper_case_globals)]
+        #[allow(unused)]
+        pub const $field: Field<$valtype, $reg_desc> =
+            Field::<$valtype, $reg_desc>::new($crate::bitmask!($numbits), $offset);
+
+        #[allow(non_snake_case)]
+        #[allow(unused)]
+        $(#[$outer])*
+        pub mod $field {
+            #[allow(unused_imports)]
+            use $crate::registers::{FieldValue, TryFromValue};
+            use super::$reg_desc;
+
+            $(
+            #[allow(non_upper_case_globals)]
+            #[allow(unused)]
+            $(#[$inner])*
+            pub const $valname: FieldValue<$valtype, $reg_desc> =
+                FieldValue::<$valtype, $reg_desc>::new($crate::bitmask!($numbits),
+                    $offset, $value);
+            )*
+
+            #[allow(non_upper_case_globals)]
+            #[allow(unused)]
+            pub const SET: FieldValue<$valtype, $reg_desc> =
+                FieldValue::<$valtype, $reg_desc>::new($crate::bitmask!($numbits),
+                    $offset, $crate::bitmask!($numbits));
+
+            #[allow(non_upper_case_globals)]
+            #[allow(unused)]
+            pub const CLEAR: FieldValue<$valtype, $reg_desc> =
+                FieldValue::<$valtype, $reg_desc>::new($crate::bitmask!($numbits),
+                    $offset, 0);
+
+            #[allow(dead_code)]
+            #[allow(non_camel_case_types)]
+            #[repr($valtype)] // so that values larger than isize::MAX can be stored
+            $(#[$outer])*
+            pub enum Value {
+                $(
+                    $(#[$inner])*
+                    $valname = $value,
+                )*
+            }
+
+            impl TryFromValue<$valtype> for Value {
+                type EnumType = Value;
+
+                fn try_from(v: $valtype) -> Option<Self::EnumType> {
+                    match v {
+                        $(
+                            $(#[$inner])*
+                            x if x == Value::$valname as $valtype => Some(Value::$valname),
+                        )*
+
+                        _ => Option::None
+                    }
+                }
+            }
+        }
+    };
+    {
+        $valtype:ident, $reg_desc:ident, $(#[$outer:meta])* $field:ident,
                     $offset:expr, $numbits:expr,
                     [$( $(#[$inner:meta])* $valname:ident = $value:expr ),*]
```

Comment:
If this is really only for the case where there are zero elements in the array, why use `*` which specifies zero or more elements rather than having:

```
{ 
$valtype:ident, $reg_desc:ident, $(#[$outer:meta])* $field:ident,
                    $offset:expr, $numbits:expr,
                    []
}
```

The answer might be that my version straight up doesn't work for some reason.  I don't really understand macros very well. Right now it seems that you're relying on the macro_rules! to match with the prior case for anything with 1 or more elements which need the `repr`, which I suppose is how it works, but feels implicit.

Reply:
using exactly that does not compile, for it to compile you also have to remove everything that uses `$inner`, `$valname` or `$value` within the portion of the macro executed by that match arm. Which I believe is actually the more correct approach, and doing so does make the macro compile, but I also am not on expert on macros and was a little worried that by changing more I would be more likely to break something.

Reply:
I think making the change to remove `$inner`, `$valname`, and `$value` as well looks pretty straightforward. I'm looking at this code and it looks like the kind of situation where either it will all explode and not compile at all, or else all work.

Reply:
Feel free to push back though.

Reply:
went ahead and took that code out and everything still compiles -- I can't think of any reason the behavior should be different